src/health_misinfo_shared/fine_tuning.py

```diff
@@ -307,26 +311,50 @@ def get_video_responses(
             generative_models.HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: generative_models.HarmBlockThreshold.BLOCK_ONLY_HIGH,
             generative_models.HarmCategory.HARM_CATEGORY_HARASSMENT: generative_models.HarmBlockThreshold.BLOCK_ONLY_HIGH,
         }
-        if len(in_context_examples) == 0:
-            response = model.predict(prompt, **parameters)
-        else:
-            response = model.generate_content(
-                [prompt],
-                generation_config=parameters,
-                safety_settings=safety_settings,
-            )
-        for candidate in response.candidates:
+        attempts = 0
+        success = False
+        while attempts < 3 and success == False:
```

Comment:
An alternative way to do this sort of thing would be:
```python
attempts = 3

for _ in range(attempts):
    # run some code here
    if success_criteria_met:
        break
else:
    # we’ve exhausted all attempts without success
    # so run the fallback code
    ...
```

This is @JamesMcMinn’s favourite `for` / `else` pattern. Once you start using it, it’s difficult to stop!

Reply:
This is a bit neater -- trying it now!